app/models/document/paginated_history.rb

```diff
@@ -0,0 +1,56 @@
+class Document::PaginatedHistory
+  PER_PAGE = 30
+
+  attr_reader :document, :query
+
+  delegate :total_count, to: :query
+
+  def initialize(document, page)
+    @document = document
+    @query = document.edition_versions
+                     .includes(:user)
+                     .where.not(state: :superseded)
+                     .reorder(created_at: :desc, id: :desc)
+                     .page(page)
+                     .per(PER_PAGE)
+  end
+
+  def audit_trail
+    query.map.with_index do |version, index|
+      # [index - 1] returns next version, as array has newest first
+      AuditTrailEntry.new(version,
+                          is_first_edition: version.item_id == document.first_edition&.id,
+                          previous_version: query[index + 1])
+    end
+  end
+
+  class AuditTrailEntry
+    attr_reader :version, :is_first_edition, :preloaded_previous_version
+
+    delegate :created_at, to: :version
+
+    def initialize(version, is_first_edition:, previous_version: nil)
+      @version = version
+      @is_first_edition = is_first_edition
+      @preloaded_previous_version = previous_version
+    end
+
+    def actor
+      version.user
+    end
+
+    def previous_version
+      # we can avoid n+1 queries by using our preloaded_prev_version
+      @previous_version ||= preloaded_previous_version || version.previous
+    end
```

Comment:
Likewise for this method. This method is only used internally within this class (to determine what the `action` should be).

Therefore this could become a private method. The only thing we really care about testing is whether the `action` is correct or not – the implementation details of `previous_version` and `preloaded_previous_version` aren't of interest to external consumers of this class.

Reply:
Done